chain/postgresdb/database/database.go

```diff
@@ -0,0 +1,334 @@
+// Package database provides support for access the database.
+package database
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/jmoiron/sqlx"
+	"github.com/lib/pq" // Calls init function.
+
+	"github.com/drand/drand/log"
+)
+
+// lib/pq errorCodeNames
+// https://github.com/lib/pq/blob/master/error.go#L178
+const (
+	uniqueViolation = "23505"
+	undefinedTable  = "42P01"
+)
+
+// Set of error variables for CRUD operations.
+var (
+	ErrDBNotFound        = sql.ErrNoRows
+	ErrDBDuplicatedEntry = errors.New("duplicated entry")
+	ErrUndefinedTable    = errors.New("undefined table")
+)
+
+// Config is the required properties to use the database.
+type Config struct {
+	User         string
+	Password     string
+	Host         string
+	Name         string
+	Schema       string
+	MaxIdleConns int
+	MaxOpenConns int
+	DisableTLS   bool
+}
+
+// ConfigFromDSN provides support for creating a config from a DSN.
+func ConfigFromDSN(dsn string) (Config, error) {
+	conf, err := url.Parse(dsn)
+	if err != nil {
+		return Config{}, err
+	}
+	query := conf.Query()
+
+	password, _ := conf.User.Password()
+	cfg := Config{
+		User:       conf.User.Username(),
+		Password:   password,
+		Host:       conf.Host,
+		Name:       conf.Path,
+		DisableTLS: true,
+	}
+
+	if query.Has("sslmode") {
+		sslMode := query.Get("sslmode")
+		switch sslMode {
+		//nolint:goconst // Having to add constants is overkill.
+		case "disable":
+			cfg.DisableTLS = true
+		case "required":
+			cfg.DisableTLS = false
+		default:
+			return Config{}, fmt.Errorf("unsupported ssl mode %q. Expected disable or required", sslMode)
+		}
+	}
+
+	cfg.MaxIdleConns = 2
+	if query.Has("max-idle") {
+		max := query.Get("max-idle")
+		m, err := strconv.Atoi(max)
```

Comment:
could somebody pass negative numbers for this? would it even matter?

Reply:
I guess that could happen. Shouldn't be a problem, but I'll force anything below 0 to be discarded.

Reply:
I would not add extra code for this check. I rather the system fail instead of adding extra code. Now you need to decide on a value to use.